layout: Support `content: <image>` on non-pseudo-elements

andreubotella opened the pull request:
Per CSS-CONTENT-3, one of the possible values of the `content` CSS property is `<content-replacement>`, which evaluates to a single `<image>`. This value is also allowed on regular elements, not just on pseudo-elements, and it will make the element into a replaced element representing the given image, discarding its contents.

This patch implements this in `traverse_element`: if the `display` value is not `none` or `contents`, we first check whether the `contents` property should make the element replaced, and if it shouldn't, then we check whether the element itself is replaced or a widget.

Per the spec, an invalid image must be treated as representing a transparent black image with zero natural width and height – in particular, it must not show a broken image icon. We added the method `ReplacedContents::zero_sized_invalid_image` to implement this.

This patch adds support for image URL references, but not for color gradients, which are treated as invalid images. The reason for this is that currently Servo does not support gradients in `ReplacedContentKind`. This is left as a follow-up change.

Testing: Some of the existing `css/css-content/element-replacement*` WPT tests now pass with this patch. We also added some new ones dealing with replacing the document root.

Fixes: #41479

andreubotella committed:
layout: Support `content: <image>` on non-pseudo-elements

Per CSS-CONTENT-3, one of the possible values of the `content` CSS
property is `<content-replacement>`, which evaluates to a single
`<image>`. This value is also allowed on regular elements, not just on
pseudo-elements, and it will make the element into a replaced element
representing the given image, discarding its contents.

This patch implements this in `traverse_element`: if the `display`
value is not `none` or `contents`, we first check whether the
`contents` property should make the element replaced, and if it
shouldn't, then we check whether the element itself is replaced or a
widget.

Per the spec, an invalid image must be treated as representing a
transparent black image with zero natural width and height – in
particular, it must not show a broken image icon. We added the method
`ReplacedContents::zero_sized_invalid_image` to implement this.

This patch adds support for image URL references, but not for color
gradients, which are treated as invalid images. The reason for this is
that currently Servo does not support gradients in
`ReplacedContentKind`. This is left as a follow-up change.

Testing: Existing `css/css-content/element-replacement*` WPT tests
that now pass with this patch. `element-replacement-gradient.html`
still fails because this patch does not add support for gradients.

Fixes: #41479

Signed-off-by: Andreu Botella <andreu@andreubotella.com>

mrobinson reviewed (commented):
Great stuff. Question below:

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#56912).

andreubotella commented:
For the record, `/fetch/metadata/generated/css-images.sub.tentative.html` and `/fetch/metadata/generated/css-images.https.sub.tentative.html` seemed to be deterministically failing in all subtests except for the `background-image` ones which were timing out, but this patch makes the `content` subtests flaky in whether they pass or fail. As far as I can tell, the reason for this is that the tests seem to expect the `content` property replacement to block the document load, even though it's not specified to do so. I'm not sure whether it blocks it in other browsers, but Firefox, Webkit and Chromium do pass these tests.

This PR also makes `html/rendering/replaced-elements/images/input-image-content.html` fail when it was previously passing, but this doesn't mean any functionality broke – it was previously passing because the reference used `<input type=image>`, which doesn't yet render the image in Servo.

I've also added new tests for the document root.

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#56912) title and body.

andreubotella commented:
> I suspect that pending images whether loaded via the image tag of CSS content should delay screenshot taking until their load completes successfully or fails.

True, but `/fetch/metadata/generated/css-images.sub.tentative.html` is not a reftest, but a testharness test that waits until an iframe's `load` event fires, and then checks which requests were made. This is why I wonder if `content` (and `background-image`, `border-image` and the various other ways of fetching images via CSS that the test is testing) are delaying the load event in other browsers, even when there's nothing in the specs to indicate that.

andreubotella commented:
I noticed that changes of the `content` property in non-pseudos did not invalidate the box tree, so I added a condition to check this when computing the layout damage.

This should have been caught by the `css/css-content/element-replacement-dynamic.html` WPT, but this test was not properly doing a layout before changing the `content` property. I changed the test to use `reftest-wait` instead.

mrobinson reviewed (approved):
Seems reasonable. I have a few nits below, but maybe @Loirooriol wants to take a look as well.

andreubotella commented on components/layout/dom_traversal.rs:
```diff
@@ -291,6 +293,22 @@ impl Contents {
+        // If `content` is a single image URL, the box gets replaced with a
+        // replaced image.
+        if let Content::Items(GenericContentItems { items, .. }) =
+            node.style(&context.style_context).clone_content()
```
Looks like Tab Atkins thinks `content` should replace other replaced elements too (https://github.com/w3c/csswg-drafts/issues/2831#issuecomment-415088965). In fact, the WPT test [`css/css-content/element-replacement-on-replaced-element.tentative.html`](https://github.com/web-platform-tests/wpt/blob/master/css/css-content/element-replacement-on-replaced-element.tentative.html) does test that `content` on a replaced image is rendered the same way as `content` on a `<video>`.

Apparently Firefox used to ignore `content` on replaced images as well, so they used to pass that test, but that ran into compat issues so they had to support it on images as well (https://bugzilla.mozilla.org/show_bug.cgi?id=1484928, https://bugzilla.mozilla.org/show_bug.cgi?id=1742411). But nowadays all major browsers support `content` on images but on no other replaced element, so they all fail that test.

I'll follow the other browsers here for now, but we should probably bring that up to the CSSWG.

andreubotella commented on components/layout/dom_traversal.rs:
```diff
@@ -291,6 +293,22 @@ impl Contents {
+                    None => {
+                        return Self::Replaced(ReplacedContents::zero_sized_invalid_image(node));
+                    },
+                }
```
Made this change in `ReplacedContents::for_element` after moving that code there.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#56912).

Loirooriol commented on components/layout/replaced.rs:
```diff
@@ -212,7 +216,7 @@ impl ReplacedContents {
             } else {
                 let element = node.as_html_element()?;
```
Non-html elements should apparently be covered too:

```html
<!DOCTYPE html>
<style>bar { content: url("https://software.hixie.ch/utilities/js/live-dom-viewer/image") }</style>
<body>
<script>
var el = document.createElementNS("foo", "bar");
el.textContent = "text";
document.body.appendChild(el);
</script>
```

andreubotella commented on components/layout/replaced.rs:
Yeah, you're right. In fact, images loaded via `content` on a pseudo-element are not animated when they should be, so this should be added to either `ReplacedContents::from_image` or `from_image_url`.
